File: ios/ClaviKeyboard/ClaviKeyboardExtension/EmojiData.swift
import Foundation

/**
 * Emoji catalogue — mirrors EmojiData.kt.
 * Loads emoji/emoji.tsv from the extension bundle once, lazily.
 */
final class EmojiData {

    struct Entry {
        let emoji: String
        let keywords: [String]
    }

    static let shared = EmojiData()

    private(set) var entries: [Entry] = []
    private var isLoaded = false

    var all: [String] { entries.map { $0.emoji } }

    func load() {
        guard !isLoaded else { return }
        isLoaded = true
        guard let url = Bundle.main.url(forResource: "emoji", withExtension: "tsv",
                                        subdirectory: "emoji"),
              let content = try? String(contentsOf: url, encoding: .utf8) else { return }
        for line in content.components(separatedBy: "\n") {
            guard !line.hasPrefix("#"), !line.isEmpty else { continue }
            guard let tabIdx = line.firstIndex(of: "\t") else { continue }
            let emoji = String(line[..<tabIdx]).trimmingCharacters(in: .whitespaces)
            let kws = String(line[line.index(after: tabIdx)...])
                .trimmingCharacters(in: .whitespaces)
                .components(separatedBy: " ")
                .filter { !$0.isEmpty }
            if !emoji.isEmpty { entries.append(Entry(emoji: emoji, keywords: kws)) }
        }
    }

    func search(_ query: String) -> [String] {
        if query.isEmpty { return all }
        let q = query.lowercased()
        return entries.filter { e in e.keywords.contains { $0.hasPrefix(q) } }.map { $0.emoji }
    }
}

File: android/app/src/main/java/com/clavi/keyboard/EmojiData.kt
package com.clavi.keyboard

import android.content.Context
import java.io.IOException

/**
 * Emoji catalogue — loads emoji.tsv from assets once, lazily.
 *
 * TSV format: emoji<TAB>keyword1 keyword2 keyword3
 * Empty line / line starting with '#' = skipped.
 *
 * Provides:
 *   - [all]         full list in file order
 *   - [search(q)]   filter by keyword prefix
 */
object EmojiData {

    data class Entry(val emoji: String, val keywords: List<String>)

    private var loaded = false
    private val entries = mutableListOf<Entry>()

    val all: List<String> get() { return entries.map { it.emoji } }

    fun load(context: Context) {
        if (loaded) return
        try {
            context.assets.open("emoji/emoji.tsv").bufferedReader().lineSequence()
                .filter { it.isNotBlank() && !it.startsWith('#') }
                .forEach { line ->
                    val tab = line.indexOf('\t')
                    if (tab < 0) return@forEach
                    val emoji = line.substring(0, tab).trim()
                    val kws = line.substring(tab + 1).trim().split(' ').filter { it.isNotBlank() }
                    if (emoji.isNotEmpty()) entries.add(Entry(emoji, kws))
                }
        } catch (_: IOException) { /* graceful degradation */ }
        loaded = true
    }

    fun search(query: String): List<String> {
        if (query.isBlank()) return all
        val q = query.trim().lowercase()
        return entries.filter { e ->
            e.keywords.any { kw -> kw.startsWith(q) }
        }.map { it.emoji }
    }
}
